### ISSUE TITLE
There is more than one instance of a concept with id {} defined in experimental sets. Please, make sure to reuse the same instance of concept

### ISSUE THREAD

**Issue by DarshanaAbeyrathna** [pytorch/captum#1164]
## To Reproduce

```
from captum.attr import LayerGradientXActivation, LayerIntegratedGradients

from captum.concept import TCAV
from captum.concept import Concept

from captum.concept._utils.data_iterator import dataset_to_dataloader, CustomIterableDataset
from captum.concept._utils.common import concepts_to_str
import glob
```
-------------------------------------------------
```
def transform(img):

    return transforms.Compose(
        [
            transforms.Resize(256),
            transforms.CenterCrop(224),
            transforms.ToTensor(),
            transforms.Normalize(
                mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]
            ),
        ]
    )(img)

def get_tensor_from_filename(filename):
    img = Image.open(filename).convert("RGB")
    return transform(img)

def load_image_tensors(class_name, root_path, transform=True):
    path = os.path.join(root_path, class_name)
    filenames = glob.glob(path + '/*.jpg')

    tensors = []
    for filename in filenames:
        img = Image.open(filename).convert('RGB')
        tensors.append(transform(img) if transform else img)
    return tensors

def assemble_concept(name, id, concepts_path):
    concept_path = os.path.join(concepts_path, name) + "/"
    dataset = CustomIterableDataset(get_tensor_from_filename, concept_path)
    concept_iter = dataset_to_dataloader(dataset)
    return Concept(id=id, name=name, data_iter=concept_iter)
```
------------------------------------------------------

```
layers=['conv_block3', 'conv_block4', 'conv_block5']

mytcav = TCAV(model=model,
              layers=layers,
              layer_attr_method = LayerIntegratedGradients(
                model, None, multiply_by_inputs=False))
```
------------------------------------------------------

```
img_path = r'C:\Users\DA\images'

kayak_imgs = load_image_tensors('kayak', img_path, transform=False)
sailboat_motor_imgs = load_image_tensors('sailboat_motor', img_path, transform=False)
```
------------------------------------------------------

```
concepts_path = r'C:\Users\DA\concept_folder'

motor_vessel_concept = assemble_concept("motor_vessel", 1, concepts_path=concepts_path)
sailboat_sail_concept = assemble_concept("sailboat_sail", 2, concepts_path=concepts_path)
```
-----------------------------------------------------

```
experimental_set = [[motor_vessel_concept, sailboat_sail_concept]]
-----------------------------------------------------

```
kayak_tensors = torch.stack([transform(img) for img in kayak_imgs])

tcav_scores_kayak_with_motor_vessel_and_sailboat_sail = mytcav.interpret(inputs=kayak_tensors,
                                         experimental_sets=experimental_set,
                                         target=0,
                                         n_steps=1)
```
------------------------------------------------------

## Expected behavior

<!-- As we can find here, I want different scores for different concepts [https://captum.ai/tutorials/TCAV_Image](https://captum.ai/tutorials/TCAV_Image). -->

**Comment by DarshanaAbeyrathna** [comment]
@orionr 
@NarineK

**Comment by orionr** [comment]
@vivekmig who on the current team could comment on this? Thank you!

**Comment by DarshanaAbeyrathna** [comment]
Hi Orion,

It seems no one going to respond? 😊

--------------------------------------
Thank you!

Best regards,
Darshana Abeyrathna Ph.D

Senior Researcher
Emerging Digital Technologies
DNV, Norway
+47 96 74 65 38
***@***.***

**Comment by NarineK** [comment]
Hi @DarshanaAbeyrathna, sorry for the delay in responding to you. I tried your code when calling `compute_cavs` method and it passed the check that you mentioned is failing. This one seems to pass:
https://github.com/pytorch/captum/blob/master/captum/concept/_core/tcav.py#L494

Do you have code stack trace of the error so that we can reproduce it ?

**Comment by DarshanaAbeyrathna** [comment]
Hi @NarineK,
When I try this with GoogleNet model:

```
model = torchvision.models.googlenet(pretrained=True)
model = model.eval()

layers=['inception4c', 'inception4d', 'inception4e']

mytcav = TCAV(model=model,
              layers=layers,
              layer_attr_method = LayerIntegratedGradients(
                model, None, multiply_by_inputs=False))
```
It works.

Previously I had my own model:

```
from torchvision import transforms

from torch.utils.data import Dataset
#from torchvision import datasets
from torchvision.transforms import ToTensor
import torchvision.models as models
from torch.utils.data import DataLoader
import torch.nn.functional as F
import torchvision

class AttentionBlock(nn.Module):
    def __init__(self, in_features_l, in_features_g, attn_features, up_factor, normalize_attn=True):
        super(AttentionBlock, self).__init__()
        self.up_factor = up_factor
        self.normalize_attn = normalize_attn
        self.W_l = nn.Conv2d(in_channels=in_features_l, out_channels=attn_features, kernel_size=1, padding=0, bias=False)
        self.W_g = nn.Conv2d(in_channels=in_features_g, out_channels=attn_features, kernel_size=1, padding=0, bias=False)
        self.phi = nn.Conv2d(in_channels=attn_features, out_channels=1, kernel_size=1, padding=0, bias=True)
    def forward(self, l, g):
        N, C, W, H = l.size()
        l_ = self.W_l(l)
        g_ = self.W_g(g)
        if self.up_factor > 1:
            g_ = F.interpolate(g_, scale_factor=self.up_factor, mode='bilinear', align_corners=False)
        c = self.phi(F.relu(l_ + g_)) # batch_sizex1xWxH
        
        # compute attn map
        if self.normalize_attn:
            a = F.softmax(c.view(N,1,-1), dim=2).view(N,1,W,H)
        else:
            a = torch.sigmoid(c)
        # re-weight the local feature
        f = torch.mul(a.expand_as(l), l) # batch_sizexCxWxH
        if self.normalize_attn:
            output = f.view(N,C,-1).sum(dim=2) # weighted sum
        else:
            output = F.adaptive_avg_pool2d(f, (1,1)).view(N,C) # global average pooling
        return a, output

class AttnVGG(nn.Module):
    def __init__(self, num_classes, normalize_attn=False, dropout=None):
        super(AttnVGG, self).__init__()
        net = models.vgg16_bn(pretrained=True)
        self.conv_block1 = nn.Sequential(*list(net.features.children())[0:6])
        self.conv_block2 = nn.Sequential(*list(net.features.children())[7:13])
        self.conv_block3 = nn.Sequential(*list(net.features.children())[14:23])
        self.conv_block4 = nn.Sequential(*list(net.features.children())[24:33])
        self.conv_block5 = nn.Sequential(*list(net.features.children())[34:43])
        self.pool = nn.AvgPool2d(7, stride=1)
        self.dpt = None
        if dropout is not None:
            self.dpt = nn.Dropout(dropout)
        self.cls = nn.Linear(in_features=512+512+256, out_features=num_classes, bias=True)
        
       # initialize the attention blocks defined above
        self.attn1 = AttentionBlock(256, 512, 256, 4, normalize_attn=normalize_attn)
        self.attn2 = AttentionBlock(512, 512, 256, 2, normalize_attn=normalize_attn)        
       
        self.reset_parameters(self.cls)
        self.reset_parameters(self.attn1)
        self.reset_parameters(self.attn2)
    def reset_parameters(self, module):
        for m in module.modules():
            if isinstance(m, nn.Conv2d):
                nn.init.kaiming_normal_(m.weight, mode='fan_in', nonlinearity='relu')
                if m.bias is not None:
                    nn.init.constant_(m.bias, 0.)
            elif isinstance(m, nn.BatchNorm2d):
                nn.init.constant_(m.weight, 1.)
                nn.init.constant_(m.bias, 0.)
            elif isinstance(m, nn.Linear):
                nn.init.normal_(m.weight, 0., 0.01)
                nn.init.constant_(m.bias, 0.)
    def forward(self, x):
        block1 = self.conv_block1(x)       # /1
        pool1 = F.max_pool2d(block1, 2, 2) # /2
        block2 = self.conv_block2(pool1)   # /2
        pool2 = F.max_pool2d(block2, 2, 2) # /4
        block3 = self.conv_block3(pool2)   # /4
        pool3 = F.max_pool2d(block3, 2, 2) # /8
        block4 = self.conv_block4(pool3)   # /8
        pool4 = F.max_pool2d(block4, 2, 2) # /16
        block5 = self.conv_block5(pool4)   # /16
        pool5 = F.max_pool2d(block5, 2, 2) # /32
        N, __, __, __ = pool5.size()
        
        g = self.pool(pool5).view(N,512)
        a1, g1 = self.attn1(pool3, pool5)
        a2, g2 = self.attn2(pool4, pool5)
        g_hat = torch.cat((g,g1,g2), dim=1) # batch_size x C
        if self.dpt is not None:
            g_hat = self.dpt(g_hat)
        out = self.cls(g_hat)

        return [out, a1, a2]

model = AttnVGG(num_classes=1, normalize_attn=True)
model=model.cpu()
```

And called it with layer names:
```
layers=['conv_block3', 'conv_block4', 'conv_block5']

mytcav = TCAV(model=model,
              layers=layers,
              layer_attr_method = LayerIntegratedGradients(
                model, None, multiply_by_inputs=False))
```

Do you think it has something to do with the model?

**Comment by NarineK** [comment]
@DarshanaAbeyrathna that's interesting because that check is independent of the model and it runs in the beginning of the interpret method call. 
https://github.com/pytorch/captum/blob/master/captum/concept/_core/tcav.py#L662
Can you still reproduce it ? If you sent us the stack trace of the error we can debug it.

**Comment by craymichael** [comment]
Closing due to inactivity.
